tron/core/actionrun.py

```diff
@@ -279,32 +302,36 @@ class ActionRun(Observable):
     # TODO: create a class for ActionRunId, JobRunId, Etc
     def __init__(
         self,
-        job_run_id,
-        name,
-        node,
-        command_config,
-        parent_context=None,
-        output_path=None,
-        cleanup=False,
-        start_time=None,
-        end_time=None,
-        run_state=SCHEDULED,
-        exit_status=None,
-        attempts=None,
-        action_runner=None,
-        retries_remaining=None,
-        retries_delay=None,
-        machine=None,
-        executor=None,
-        trigger_downstreams=None,
-        triggered_by=None,
-        on_upstream_rerun=None,
-        trigger_timeout_timestamp=None,
-        original_command=None,
+        job_run_id: str,
+        name: str,
+        node: node.Node,
+        command_config: action.ActionCommandConfig,
+        parent_context: Optional[CommandContext] = None,
+        output_path: Optional[filehandler.OutputPath] = None,
+        cleanup: bool = False,
+        start_time: Optional[datetime.datetime] = None,
+        end_time: Optional[datetime.datetime] = None,
+        run_state: str = SCHEDULED,
+        exit_status: Optional[int] = None,
+        attempts: Optional[list] = None,  # TODO: list of...ActionCommandConfig?
+        action_runner: Optional[Union[NoActionRunnerFactory, SubprocessActionRunnerFactory]] = None,
```

Comment:
maybe one day we'll either remove `NoActionRunnerFactory` (i don't think we use it outside of tests) or we'll add an abstract base class or something with the expected interface so that we don't need a union here :p

Reply:
Added a ticket for some noble soul in a future unhack 😛